pkg/gadgets/profile/tracer/core/tracer.go

```diff
@@ -0,0 +1,388 @@
+//go:build linux
+// +build linux
+
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracer
+
+// #include <linux/types.h>
+// #include "./bpf/profile.h"
+import "C"
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"runtime"
+	"sort"
+	"strconv"
+	"strings"
+	"unsafe"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/cilium/ebpf"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/profile/tracer"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/profile/types"
+	"golang.org/x/sys/unix"
+)
+
+//go:generate sh -c "GOOS=$(go env GOHOSTOS) GOARCH=$(go env GOHOSTARCH) go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang profile ./bpf/profile.bpf.c -- -I./bpf/ -I../../../../ -target bpf -D__TARGET_ARCH_x86"
+
+type Tracer struct {
+	resolver gadgets.Resolver
+	node     string
+	objs     profileObjects
+	perfFds  []int
+	config   *tracer.Config
+}
+
+const (
+	perfMaxStackDepth = 127
+	perfSampleFreq    = 49
+	// In C, struct perf_event_attr has a freq field which is a bit in a
+	// 64-length bitfield.
+	// In Golang, there is a Bits field which 64 bits long.
+	// From C, we can deduce freq (which permits using frequency not period)
+	// is the 10th bit.
+	frequencyBit = 1 << 10
+)
+
+func NewTracer(resolver gadgets.Resolver, config *tracer.Config, node string) (*Tracer, error) {
+	t := &Tracer{
+		resolver: resolver,
+		node:     node,
+		config:   config,
+	}
+
+	if err := t.start(); err != nil {
+		t.Stop()
+		return nil, err
+	}
+
+	return t, nil
+}
+
+type keyCount struct {
+	key   C.struct_key_t
+	value uint64
+}
+
+func readCountsMap(counts *ebpf.Map) ([]keyCount, error) {
```

Comment:
nit: why not to make this method part of the Tracer structure to avoid having to pass the ebpf map?